### PR TITLE
`class << self` style syntax support

### PR DESCRIPTION
**Pull request by telyn**
This is a first-pass at implementing #258

I haven't finished writing tests for it (it's been such an experimental process trying to figure out what needs to go where that TDD quickly became too difficult) and my local is currently still full of `puts` and `pry` - so I'll need to go over this diff with a fine-toothed comb to make sure I didn't leak any kind of debugging stuff, and find what I've not tested and write tests for it. P.s. I'm not entirely expecting the tests to pass as they are either.

I'm submitting this PR now as a draft to gain feedback - does it seem like I have forgotten something or done too much somewhere, or violated any principles of style or responsibility? Are my names bad? I think my names might be bad.

*Note to self*:
* [x] Fix existing tests
* [ ] Finish new tests
* [x] Fix the SingletonMetaclass / Metaclass naming inconsistency

### CODE REVIEW

**Comment by mbj** [comment]
@telyn First thank you for this PR.

Just a "word of warning". Mutant is now commercial software. Its as of release >=0.9 not OSS anymore. If you are still comfortable to contribute under that new arrangement (I had to make as maintaining mutant was too much work on OSS economy): I'd have to ask you to sign a CLA for getting this PR merged.

Other than that I'm happy to review, but I'll not be able to merge without signing the CLA.

If you wanted to continue with the CLA please contact me on my email with the subject `Mutant CLA`.

Other than that: I should be able to give feedback this week. If you do not hear anything: Make sure you *ping me here and bug me*.

**Comment by telyn** [comment]
Thanks for the speedy reply!

I knew it was commercial yeah, figured there might be some process to it. Worth it for a project this cool, ha :)

Will bug you if I remember to! 😅

**Comment by mbj** [comment]
@telyn I just gave this a small birds eye pass, not flagging individual instances.

Overall you seem to follow the existing style of the code, which is much appreciated.

You correctly spotted to make a meta/eigenclass specific subject to emit the monkeypatch correctly. Thanks for that. This keeps the constant scope etc equivalent.

Overall you are on the right track, once you have it green I'm happy to do a full review. Just ping me.

Note that I have to insist on the CLA signature before I'll be able to merge. You should have it in your inbox right now.

**Comment by telyn** [comment]
Thanks for that Markus - will get the CLA to you and get this green hopefully this weekend.

**Comment by mbj** [comment]
@telyn You'll notice that I fixed the build meanwhile, so you should be able to rebase and get a green build.

**Comment by telyn** [comment]
> @telyn You'll notice that I fixed the build meanwhile, so you should be able to rebase and get a green build.

Thanks! I've got some more tests to write, then I'll mark the PR as ready for review :-)

**Comment by telyn** [comment]
Just to let you know I've not forgotten about this - I've just been feeling __super terrible__ lately and barely been feeling up to it. 

The good news is I'm pretty sure it's just a collection of methods I refactored into a class that need unit tests. So it's a relatively easy job once I get back to it :-)

**Comment by mbj** [comment]
> Just to let you know I've not forgotten about this - I've just been feeling **super terrible** lately and barely been feeling up to it.
> 
> The good news is I'm pretty sure it's just a collection of methods I refactored into a class that need unit tests. So it's a relatively easy job once I get back to it :-)

No worries. This is opensource, nobody expects you to harm yourself in doing this kind of contrubtions.

**Comment by telyn** [comment]
Hey Martin - I think this is ready for a complete review now :star: 

I'm sure I've got some dodgy naming in there (`Metaclass`? `Eigenclass`? `SingletonClass`? decisions are hard) - very happy for you to want changes!

**Comment by mbj** [comment]
> Hey Martin - I think this is ready for a complete review now 

Its Markus, but no worries. We also have a Martin here @snusnu  ;)

> I'm sure I've got some dodgy naming in there (`Metaclass`? `Eigenclass`? `SingletonClass`? decisions are hard) - very happy for you to want changes!

I think all of them have merit, and I do not have an off hand preference. Will have to do some research to determine which one is most appropriate. Probably looking at MRI and how they name these things internally.

On the review: I'm to busy to jump on it this week. Can you make sure you ping me in case I drag this out too long?

**Comment by telyn** [comment]
> > Hey Martin - I think this is ready for a complete review now
> 
> Its Markus, but no worries. We also have a Martin here @snusnu ;)
> 
> > I'm sure I've got some dodgy naming in there (`Metaclass`? `Eigenclass`? `SingletonClass`? decisions are hard) - very happy for you to want changes!
> 
> I think all of them have merit, and I do not have an off hand preference. Will have to do some research to determine which one is most appropriate. Probably looking at MRI and how they name these things internally.
> 
> On the review: I'm to busy to jump on it this week. Can you make sure you ping me in case I drag this out too long?

Hey again - sorry for getting your name wrong >.< Complete rookie mistake!

I also totally forgot to poke you to make sure this got reviewed and merged - sorry about that. Memory of a goldfish, me.

Consider this a reminder (albeit a late one!)

File: lib/mutant/matcher/method/singleton.rb [hunk 1]
```diff
@@ -20,7 +20,8 @@ def self.new(scope, method_name)
         class Evaluator < Evaluator
           SUBJECT_CLASS    = Subject::Method::Singleton
           RECEIVER_INDEX   = 0
-          NAME_INDEX       = 1
+          DEFS_NAME_INDEX  = 1
+          DEF_NAME_INDEX    = 0
```
Comment (mbj) [inline lib/mutant/matcher/method/singleton.rb]: whitespace not aligned.

File: spec/unit/mutant/matcher/method/singleton_spec.rb [hunk 2]
```diff
@@ -57,7 +57,15 @@ def arguments
 
   context 'when defined on metaclass' do
     let(:scope)       { base::DefinedOnMetaclass }
+    let(:type)        { :def }
```
Comment (mbj) [inline spec/unit/mutant/matcher/method/singleton_spec.rb]: alignment.

File: lib/mutant/matcher/method/metaclass.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            mc = metaclass(node)
+            mc && receiver?(mc)
+          end
+
+          def metaclass(node)
+            Mutant::AST.find_last_path(ast) do |cur_node|
+              next unless n_sclass?(cur_node)
+
+              cur_node.children.index do |child|
+                child.equal? node
```
Comment (mbj) [inline lib/mutant/matcher/method/metaclass.rb]: missing parens on `node` argument.

**Comment by mbj** [comment]
@telyn So I had a time to do a full pass, but stopped as the commit stream makes this really hard to review.

The commit stream is not canonical for this project. A canonical commit stream is formed of commits that have the following properties:

* Each commit falls into one of the classes of semantic: Removal, Fix, Change, Add.
* Each commit passes the CI
* Each commit could be deployed to production (released in mutants case) without causing a regression.

I think the simplest way to get to this state is if you where to squash down your work to one `Add support for metaclass` commit.

From there we may find nice extractions. The first commit likely can stay as is, its a good one.

I'm fine to do this work of making the commit stream canoncial, just tell me to.

**Comment by telyn** [comment]
Thanks for performing the review Markus :)

I'd be happy to perform that canonicalisation - I was unaware of that requirement, though I totally understand what you're saying about it being tough to bisect. I'm currently in the middle of moving house, so may have to wait until next weekend or so, if that's ok?

Cheers :-)

**Comment by mbj** [comment]
> I'd be happy to perform that canonicalisation - I was unaware of that requirement, though I totally understand what you're saying about it being tough to bisect. I'm currently in the middle of moving house, so may have to wait until next weekend or so, if that's ok?

I'll do it by myself, tonight. Do not want to wait any longer, given I'm the delay factor here.

**Comment by mbj** [comment]
@telyn I'm closing this PR as I made #1007 on this repository from collapsing your commits into one. I'll add my fixups there and collapse my fixups once done into your work. I'll keep your commit authorship intact.

**Comment by telyn** [comment]
Ace - that works excellently for me! Thanks for everything Markus :)

**Comment by mbj** [comment]
@telyn This change was released in 0.9.9.

**Pull request closed without merging**